### PR TITLE
Require that each pattern have at least one RHS.

### LINKED ISSUE #3584: Bad handling of missing arrow in pattern matching
The following code:

```
match Some 3 with
  None ->
    "none"
  Some _
    "some"
```

Produces this output:

```
💔💥
  
  I've encountered a pattern match failure in function `#d6r72knghh` while scrutinizing:
  
    Some 3
  
  This happens when calling a function that doesn't handle all possible inputs
  
  
  Stack trace:
    base.bug
    #4a7largfa5
```

The issue is that the code should have `Right _ ->` where it has `Right _`. This code definitely shouldn't compile, and it should probably be a parse error.

### PR DESCRIPTION
**Pull request by ChrisPenner**
Fixes https://github.com/unisonweb/unison/issues/3584

## Overview

The parser didn't require that your match cases have a `->`, so you could parse weird cases like:

Ex. 1
```
match Some 3 with
  None ->
    "none"
  Some _
    "some"
```

Where the parser would just throw away the everything after "none" since it doesn't have a RHS.

While looking through the parser I also discovered you can do silly things like this:

Ex. 2
```
x = match None with
      None -> 0
           -> 1
           -> 2
```

Ex. 3
```
x = match None with
      None -> 0
          | true -> 1
```

Ex. 4
```
x = match None with
      Some _
        | true -> 1
        -> 2
```

## Implementation notes

* Rewrote the pattern case parser so it _requires_ a RHS to every pattern.
* In the process, disallowed guards from following non-guards, since this is nonsensical (within a single pattern, you can still have guards on a subsequent pattern). See Ex. 3 and Ex 4. 
* Disallowed multiple non-guarded `->` blocks for a given pattern. (see Ex. 2)

## Test coverage

Update the error-messages transcript

### CODE REVIEW

File: parser-typechecker/src/Unison/Syntax/TermParser.hs [hunk 1]
```diff
@@ -161,27 +161,33 @@ matchCases1 start = do
...
       pair p1 p2 = Pattern.Constructor (ann p1 <> ann p2) (ConstructorReference DD.pairRef 0) [p1, p2]
-  guardsAndBlocks <- many $ do
-    guard <-
-      asum
-        [ Nothing <$ P.try (reserved "|" *> quasikeyword "otherwise"),
-          optional $ reserved "|" *> infixAppOrBooleanOp
-        ]
-    t <- block "->"
-    pure (guard, t)
+  let guardedBlocks = label "pattern guard" . some $ do
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Syntax/TermParser.hs:171]: We now parse guarded and unguarded blocks separately, each pattern either has one or more guards, or no guards at all, you can't mix and match guarded and unguarded RHSs.

File: parser-typechecker/src/Unison/Syntax/TermParser.hs [hunk 2]
```diff
@@ -161,27 +161,33 @@ matchCases1 start = do
...
 matchCase = do
-  pats <- sepBy1 (reserved ",") parsePattern
+  pats <- sepBy1 (label "\",\"" $ reserved ",") parsePattern
   let boundVars' = [v | (_, vs) <- pats, (_ann, v) <- vs]
       pat = case fst <$> pats of
         [p] -> p
         pats -> foldr pair (unit (ann . last $ pats)) pats
       unit ann = Pattern.Constructor ann (ConstructorReference DD.unitRef 0) []
       pair p1 p2 = Pattern.Constructor (ann p1 <> ann p2) (ConstructorReference DD.pairRef 0) [p1, p2]
-  guardsAndBlocks <- many $ do
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Syntax/TermParser.hs:171]: This `many` meant that you could get away with having a pattern that doesn't have any RHS block.

File: unison-src/transcripts/error-messages.output.md [hunk 3]
```diff
@@ -195,6 +208,56 @@ foo = cases
...
+x = match Some a with
+      None -> 
+        1
+      Some _
+        2
+```
+
+```ucm
+
+  offset=16:
```
Comment (ChrisPenner) [inline unison-src/transcripts/error-messages.output.md:223]: Anyone know what these `offsets` are in our parser error messages? Doesn't seem helpful to me 🤔

**Review by pchiusano — APPROVED** [review]

**Review by ceedubs — APPROVED** [review]
🎉 woohoo thanks, @ChrisPenner!

**Pull request merged**